change to an option for time checking storage.reserve

perrynzhou opened the pull request:
storage/posix: change to an option for time checking storage.reserve (#2120)

Change-Id: 66e34de17b1e6e308a5b648fc2232f543047b40d
Fixes: #2120
Signed-off-by: perrynzhou <perrynzhou@gmail.com>

code detail:
- https://github.com/gluster/glusterfs/commit/4c13280c813694f9eb1ad2cb1d864098ea01a006
- https://github.com/gluster/glusterfs/commit/573d133f463dd8dde5c6fad98fa91b827682c348

Commit:
change to an option for time checking storage.reserve
storage/posix: change to an option for time checking storage.reserve (#2120)

Change-Id: 66e34de17b1e6e308a5b648fc2232f543047b40d
Fixes: #2120
Signed-off-by: perrynzhou <perrynzhou@gmail.com>

gluster-ant commented:
Can one of the admins verify this patch?

gluster-ant commented:
Can one of the admins verify this patch?

gluster-ant commented:
Can one of the admins verify this patch?

pranithk commented on xlators/mgmt/glusterd/src/glusterd-volume-set.c:
```diff
@@ -2386,6 +2386,11 @@ struct volopt_map_entry glusterd_volopt_map[] = {
+    {
+        .key = "storage.reserve-check-interval",
+        .voltype = "storage/posix",
+        .op_version = GD_OP_VERSION_4_0_0,
```
This should be latest version where you can have this included. So GD_OP_VERSION_9_1

pranithk commented on xlators/mgmt/glusterd/src/glusterd-volume-set.c:
```diff
@@ -2386,6 +2386,11 @@ struct volopt_map_entry glusterd_volopt_map[] = {
+        .key = "storage.reserve-check-interval",
+        .voltype = "storage/posix",
+        .op_version = GD_OP_VERSION_4_0_0,
+    },
```
You also need to add .voltype = "storage/posix"

pranithk commented on xlators/storage/posix/src/posix-common.c:
```diff
@@ -418,6 +418,9 @@ posix_reconfigure(xlator_t *this, dict_t *options)
     old_disk_reserve = priv->disk_reserve;
     GF_OPTION_RECONF("reserve", priv->disk_reserve, options, percent_or_size,
                      out);
+    // reload  config  to reserve check  for check thread
```
remove the comment.

pranithk commented on xlators/storage/posix/src/posix-common.c:
```diff
@@ -1120,7 +1123,8 @@ posix_init(xlator_t *this)
 
     GF_OPTION_INIT("reserve", _private->disk_reserve, percent_or_size, out);
-
+    // add disk reserve internal seconds for reserve check
```
remove the comment here as well

pranithk commented on xlators/storage/posix/src/posix-common.c:
```diff
@@ -1449,6 +1453,14 @@ struct volume_options posix_options[] = {
+     .default_value = "5",
+     .validate = GF_OPT_VALIDATE_MIN,
+     .description =  "Interval in second to check disk reserve",
+     .op_version = {GD_OP_VERSION_4_0_0},
```
change version

pranithk commented on xlators/storage/posix/src/posix-helpers.c:
```diff
@@ -2361,12 +2361,6 @@ posix_ctx_disk_thread_proc(void *data)
-    gf_msg_debug("glusterfs_ctx", 0,
-                 "Ctx disk-space thread started, "
-                 "interval = %d seconds",
-                 interval);
```
Keep the log, just change the interval variable with the variable you added.

perrynzhou replied:
@pranithk  i found  need pass xlator pointer to posix_ctx_disk_thread_proc,not glusterfs_ctx_t pointer.

Commit:
change to an option for time checking storage.reserve,add GD_OP_VERSION_9_1 and remove comment
storage/posix: change to an option for time checking storage.reserve (#2120)

Change-Id: 66e34de17b1e6e308a5b648fc2232f543047b40d
Fixes: #2120
Fixed: #2120
Signed-off-by: perrynzhou <perrynzhou@gmail.com>

pranithk commented:
@perrynzhou clang format check is failing. Could you run 
```
git show --pretty="format:" --name-only | grep -v "contrib/" | egrep --color=never "*\.[ch]$" | xargs clang-format -i
```
in your repo and commit those changes also? Other changes look okay to me. Adding other maintainers also.

mohit84 commented on xlators/storage/posix/src/posix.h:
```diff
@@ -202,6 +202,8 @@ struct posix_private {
 
     double disk_reserve;
+    /* seconds for check disk reversion  */
+    uint32_t disk_reserve_check_interval;
```
I think you should declare disk_reserve_check_interval in under glusterfs_ctx_t because currently we have single posixreserve thread for all the bricks while brick_mux is enabled.If user changes the value for any volume it should reset ctx->disk_reserve_check_interval variable and use the same at the time of using the interval
in posix_ctx_disk_thread_proc.

perrynzhou replied:
i agree  @mohit84.

perrynzhou commented:
> ```
> git show --pretty="format:" --name-only | grep -v "contrib/" | egrep --color=never "*\.[ch]$" | xargs clang-format -i
> ```

@pranithk  it's invalid  .clang_format.how to fixed that?

```
[root@CentOS glusterfs]# git show --pretty="format:" --name-only | grep -v "contrib/" | egrep --color=never "*\.[ch]$" | xargs clang-format -i
YAML:38:29: error: invalid boolean
BreakBeforeBinaryOperators: None
                            ^~~~
Error reading /root/Source/zhoulin/glusterfs/.clang-format: Invalid argument
Can't find usable .clang-format, using LLVM style
YAML:38:29: error: invalid boolean
BreakBeforeBinaryOperators: None
                            ^~~~
Error reading /root/Source/zhoulin/glusterfs/.clang-format: Invalid argument
Can't find usable .clang-format, using LLVM style
YAML:38:29: error: invalid boolean
BreakBeforeBinaryOperators: None
                            ^~~~
Error reading /root/Source/zhoulin/glusterfs/.clang-format: Invalid argument
Can't find usable .clang-format, using LLVM style
YAML:38:29: error: invalid boolean
BreakBeforeBinaryOperators: None
                            ^~~~
Error reading /root/Source/zhoulin/glusterfs/.clang-format: Invalid argument
Can't find usable .clang-format, using LLVM style
```

pranithk commented:
> > ```
> > git show --pretty="format:" --name-only | grep -v "contrib/" | egrep --color=never "*\.[ch]$" | xargs clang-format -i
> > ```
> 
> @pranithk it's invalid .clang_format.how to fixed that?
> 
> ```
> [root@CentOS glusterfs]# git show --pretty="format:" --name-only | grep -v "contrib/" | egrep --color=never "*\.[ch]$" | xargs clang-format -i
> YAML:38:29: error: invalid boolean
> BreakBeforeBinaryOperators: None
>                             ^~~~
> Error reading /root/Source/zhoulin/glusterfs/.clang-format: Invalid argument
> Can't find usable .clang-format, using LLVM style
> YAML:38:29: error: invalid boolean
> BreakBeforeBinaryOperators: None
>                             ^~~~
> Error reading /root/Source/zhoulin/glusterfs/.clang-format: Invalid argument
> Can't find usable .clang-format, using LLVM style
> YAML:38:29: error: invalid boolean
> BreakBeforeBinaryOperators: None
>                             ^~~~
> Error reading /root/Source/zhoulin/glusterfs/.clang-format: Invalid argument
> Can't find usable .clang-format, using LLVM style
> YAML:38:29: error: invalid boolean
> BreakBeforeBinaryOperators: None
>                             ^~~~
> Error reading /root/Source/zhoulin/glusterfs/.clang-format: Invalid argument
> Can't find usable .clang-format, using LLVM style
> ```

What is your clang-format version?

perrynzhou commented:
@pranithk  `LLVM version 3.4.2` for clang-format

```
[root@CentOS glusterfs]# clang-format --version
LLVM (http://llvm.org/):
  LLVM version 3.4.2
  Optimized build.
  Built Nov  1 2018 (15:06:24).
  Default target: x86_64-redhat-linux-gnu
  Host CPU: x86-64
```

pranithk commented:
> @pranithk `LLVM version 3.4.2` for clang-format
> 
> ```
> [root@CentOS glusterfs]# clang-format --version
> LLVM (http://llvm.org/):
>   LLVM version 3.4.2
>   Optimized build.
>   Built Nov  1 2018 (15:06:24).
>   Default target: x86_64-redhat-linux-gnu
>   Host CPU: x86-64
> ```

Why is it printing LLVM info?

This is the output on my machine:
```
19:07:29 :) ⚡ clang-format --version
clang-format version 11.0.0 (Fedora 11.0.0-2.fc33)

19:07:32 :) ⚡ which clang-format 
/usr/bin/clang-format

19:17:46 :) ⚡ ls -l /usr/bin/clang-format
-rwxr-xr-x. 1 root root 79944 Nov  3 15:05 /usr/bin/clang-format

```

perrynzhou commented:
@pranithk  i check that.i declare disk_reserve_check_interval in under glusterfs_ctx_t,when i test all,i will push later

```
[root@CentOS ~]# clang-format -version
clang-format version 5.0.1 (tags/RELEASE_501/final)
```

Commit:
declare disk_reserve_check_interval in under glusterfs_ctx_t
storage/posix: change to an option for time checking storage.reserve (#2120)

Change-Id: 66e34de17b1e6e308a5b648fc2232f543047b40d
Fixes: #2120
Signed-off-by: perrynzhou <perrynzhou@gmail.com>

perrynzhou commented:
@xhernandez  @pranithk  @amarts  @mohit84  link https://github.com/gluster/glusterfs/pull/2150 , pr https://github.com/gluster/glusterfs/pull/2150/commits/4c13280c813694f9eb1ad2cb1d864098ea01a006 ,please review all

perrynzhou commented:
@pranithk @xhernandez @amarts link for pr https://github.com/gluster/glusterfs/pull/2150/commits/4c13280c813694f9eb1ad2cb1d864098ea01a006 ,please review all,thank you

pranithk commented:
> @xhernandez @pranithk @amarts @mohit84 link #2150 , pr [4c13280](https://github.com/gluster/glusterfs/commit/4c13280c813694f9eb1ad2cb1d864098ea01a006) ,please review all

Sorry, this is confusing. Why did you raise PR 2150? Was it by mistake? You can push changes to perrynzhou:perryn/dev itself for the new changes to appear in this PR.

perrynzhou commented:
@pranithk   when i run ./rfc.sh, that build  new pr

pranithk commented:
@perrynzhou Okay, please use `git push origin <branch-name>` and submit any more changes sometimes you may have to give -f option to push. I think we need to fix ./rfc.sh for submitting new changes to exisiting PRs.

perrynzhou commented:
@pranithk ,need i re-commit pr? when i push ,the logs such as
```
[root@CentOS glusterfs]# git push origin perryn/dev -f
Username for 'https://github.com': perrynzhou
Password for 'https://perrynzhou@github.com': 
Counting objects: 11, done.
Delta compression using up to 12 threads.
Compressing objects: 100% (6/6), done.
Writing objects: 100% (6/6), 699 bytes | 0 bytes/s, done.
Total 6 (delta 4), reused 0 (delta 0)
remote: Resolving deltas: 100% (4/4), completed with 4 local objects.
To https://github.com/perrynzhou/glusterfs.git
   4c13280..573d133  perryn/dev -> perryn/dev
[root@CentOS glusterfs]# ./rfc.sh 
Picked upstream as upstream remote
Picked origin as user's remote
Picked upstream as upstream remote
Picked origin as user's remote
Picked upstream as upstream remote
Picked origin as user's remote
[detached HEAD 79d05fe] change to an option for time checking storage.reserve storage/posix: change to an option for time checking storage.reserve (#2120)
 4 files changed, 21 insertions(+), 8 deletions(-)
Picked upstream as upstream remote
Picked origin as user's remote
[detached HEAD 2028164] change to an option for time checking storage.reserve,add GD_OP_VERSION_9_1 and remove comment storage/posix: change to an option for time checking storage.reserve (#2120)
 4 files changed, 12 insertions(+), 10 deletions(-)
Picked upstream as upstream remote
Picked origin as user's remote
[detached HEAD 747d833] declare disk_reserve_check_interval in under glusterfs_ctx_t storage/posix: change to an option for time checking storage.reserve (#2120)
 5 files changed, 16 insertions(+), 28 deletions(-)
Picked upstream as upstream remote
Picked origin as user's remote
[detached HEAD 27e4017] declare disk_reserve_check_interval in under glusterfs_ctx_t storage/posix: change to an option for time checking storage.reserve (#2120)
 1 file changed, 1 insertion(+)
Successfully rebased and updated refs/heads/perryn/dev.
To https://github.com/perrynzhou/glusterfs.git
 ! [rejected]        HEAD -> issue2120_devel (non-fast-forward)
error: failed to push some refs to 'https://github.com/perrynzhou/glusterfs.git'
hint: Updates were rejected because a pushed branch tip is behind its remote
hint: counterpart. Check out this branch and merge the remote changes
hint: (e.g. 'git pull') before pushing again.
hint: See the 'Note about fast-forwards' in 'git push --help' for details.
```

Commit:
declare disk_reserve_check_interval in under glusterfs_ctx_t
storage/posix: change to an option for time checking storage.reserve (#2120)

Change-Id: 66e34de17b1e6e308a5b648fc2232f543047b40d
Fixes: #2120
Signed-off-by: perrynzhou <perrynzhou@gmail.com>

pranithk commented:
/recheck smoke

pranithk commented:
/run regression

pranithk commented:
Sorry, I missed update on this PR.I triggered the regression. For some reason since yesterday there are spurious failures with builds. We may have to trigger builds multiple times. I will wait for @mohit84 review

perrynzhou commented:
@pranithk   thank you very much. owner branch named perryn/dev already automerge to glusterfs/dev. the detail code changes link such as 
 - https://github.com/gluster/glusterfs/commit/4c13280c813694f9eb1ad2cb1d864098ea01a006
-  https://github.com/gluster/glusterfs/commit/573d133f463dd8dde5c6fad98fa91b827682c348  

https://github.com/gluster/glusterfs/commit/4c13280c813694f9eb1ad2cb1d864098ea01a006 changes core code for this issue,  https://github.com/gluster/glusterfs/commit/573d133f463dd8dde5c6fad98fa91b827682c348   add comment 
i still wait for @mohit84  @amarts  @xhernandez  review

mohit84 commented:
/run brick-mux regression

mohit84 commented:
/run brick-mux regression

pranithk commented:
@mohit84 @perrynzhou If we set option on one volume, will it change behavior for all volumes?

mohit84 commented:
> @mohit84 @perrynzhou If we set option on one volume, will it change behavior for all volumes?

Yes

pranithk commented:
> > @mohit84 @perrynzhou If we set option on one volume, will it change behavior for all volumes?
> 
> Yes

That would be considered a bug.

pranithk reviewed (changes requested):
Changing status so that the patch with this behavior won't be merged before votes of others also.

There are 2 options:
1) Move this option to gluster volume set all command
2) Make it work for the volume.

I think option 2 is better. exposing posix's behavior in glusterfs_ctx seems wrong from design perspective.

mohit84 commented:
> Changing status so that the patch with this behavior won't be merged before votes of others also.
> 
> There are 2 options:
> 
> 1. Move this option to gluster volume set all command
> 2. Make it work for the volume.
> 
> I think option 2 is better. exposing posix's behavior in glusterfs_ctx seems wrong from design perspective.

It is not possible to maintain the volume specific sleep interval with current design because we have attached disk_space_thread with glusterfs_ctx. In real time environment 5 sec sleep time is sufficient to check disk space, it is not realistic to check disk space at every 1 s basis, we never faced any issue with 5 sec basis.

pranithk commented:
> > Changing status so that the patch with this behavior won't be merged before votes of others also.
> > There are 2 options:
> > 
> > 1. Move this option to gluster volume set all command
> > 2. Make it work for the volume.
> > 
> > I think option 2 is better. exposing posix's behavior in glusterfs_ctx seems wrong from design perspective.
> 
> It is not possible to maintain the volume specific sleep interval with current design because we have attached disk_space_thread with glusterfs_ctx. In real time environment 5 sec sleep time is sufficient to check disk space, it is not realistic to check disk space at every 1 s basis, we never faced any issue with 5 sec basis.

I think it depends on the workload whether 5 seconds is good or not. So it makes sense to make it a configurable value. I think the current design is limiting. Maybe the best way forward would be to get disk_space_thread to accept a queue of requests along with any other tasks it may want to do. Let posix etc register time events to push events to that queue. I think that would make this work in this scenario.

perrynzhou commented:
@mohit84  @pranithk  it's change  internal time for disk_space_thread  to option config  is better.not all need 5 seconds to check. i think we need change only  disk_reserve_check_interval for each volume not all volume.hard code is not flexible.If we set option on one volume, if it change behavior for all volumes ,that consider bug

perrynzhou commented:
@pranithk  @xhernandez  @amarts  please review all changes,thank you very much

perrynzhou commented:
@mohit84  when all changes will be merged to glusterfs?
